SPARK-12729 PhantomReferences to replace Finalize()

GayathriMurali opened the pull request:
PhantomReference to replace finalize() method in Python Broadcast.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -892,27 +911,31 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
-    }
-  }
+    }.start() 
+   
```
whitespace

holdenk commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -865,6 +869,17 @@ private class PythonAccumulatorParam(@transient private val serverHost: String,
+
+private[spark] class FilePhantomReference(@transient var f:File, var q: ReferenceQueue[File]) extends PhantomReference(f,q){
+
+	private def cleanup()
```
check indentation here

holdenk commented:
Thanks for making this PR - at first glance there are some minor styling issues that might be good to fix (see our style guide at https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide )

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

GayathriMurali commented:
Fixed styling issues. Please review

vanzin commented:
ok to test

SparkQA commented:
**[Test build #51003 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51003/consoleFull)** for PR 11140 at commit [`14876ee`](https://github.com/apache/spark/commit/14876ee2e549c998753b577e8148832dbc4822f5).

SparkQA commented:
**[Test build #51003 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51003/consoleFull)** for PR 11140 at commit [`14876ee`](https://github.com/apache/spark/commit/14876ee2e549c998753b577e8148832dbc4822f5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51003/
Test FAILed.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

GayathriMurali commented:
Recent commit fixes all the styling bugs reported.

SparkQA commented:
**[Test build #51006 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51006/consoleFull)** for PR 11140 at commit [`b6c9caf`](https://github.com/apache/spark/commit/b6c9cafa80f8bdc2540ae4e0a2172b968ad553b4).

SparkQA commented:
**[Test build #51006 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51006/consoleFull)** for PR 11140 at commit [`b6c9caf`](https://github.com/apache/spark/commit/b6c9cafa80f8bdc2540ae4e0a2172b968ad553b4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51006/
Test FAILed.

zsxwing committed:
[SPARK-13245][CORE] Call shuffleMetrics methods only in one thread for ShuffleBlockFetcherIterator

Call shuffleMetrics's incRemoteBytesRead and incRemoteBlocksFetched when polling FetchResult from `results` so as to always use shuffleMetrics in one thread.

Also fix a race condition that could cause memory leak.

Author: Shixiong Zhu <shixiong@databricks.com>

Closes #11138 from zsxwing/SPARK-13245.

Commit:
[SPARK-12950] [SQL] Improve lookup of BytesToBytesMap in aggregate

This PR improve the lookup of BytesToBytesMap by:

1. Generate code for calculate the hash code of grouping keys.

2. Do not use MemoryLocation, fetch the baseObject and offset for key and value directly (remove the indirection).

Author: Davies Liu <davies@databricks.com>

Closes #11010 from davies/gen_map.

viirya committed:
[SPARK-10524][ML] Use the soft prediction to order categories' bins

JIRA: https://issues.apache.org/jira/browse/SPARK-10524

Currently we use the hard prediction (`ImpurityCalculator.predict`) to order categories' bins. But we should use the soft prediction.

Author: Liang-Chi Hsieh <viirya@gmail.com>
Author: Liang-Chi Hsieh <viirya@appier.com>
Author: Joseph K. Bradley <joseph@databricks.com>

Closes #8734 from viirya/dt-soft-centroids.

maropu committed:
[SPARK-12476][SQL] Implement JdbcRelation#unhandledFilters for removing unnecessary Spark Filter

Input: SELECT * FROM jdbcTable WHERE col0 = 'xxx'

Current plan:
```
== Optimized Logical Plan ==
Project [col0#0,col1#1]
+- Filter (col0#0 = xxx)
   +- Relation[col0#0,col1#1] JDBCRelation(jdbc:postgresql:postgres,testRel,[Lorg.apache.spark.Partition;2ac7c683,{user=maropu, password=, driver=org.postgresql.Driver})

== Physical Plan ==
+- Filter (col0#0 = xxx)
   +- Scan JDBCRelation(jdbc:postgresql:postgres,testRel,[Lorg.apache.spark.Partition;2ac7c683,{user=maropu, password=, driver=org.postgresql.Driver})[col0#0,col1#1] PushedFilters: [EqualTo(col0,xxx)]
```

This patch enables a plan below;
```
== Optimized Logical Plan ==
Project [col0#0,col1#1]
+- Filter (col0#0 = xxx)
   +- Relation[col0#0,col1#1] JDBCRelation(jdbc:postgresql:postgres,testRel,[Lorg.apache.spark.Partition;2ac7c683,{user=maropu, password=, driver=org.postgresql.Driver})

== Physical Plan ==
Scan JDBCRelation(jdbc:postgresql:postgres,testRel,[Lorg.apache.spark.Partition;2ac7c683,{user=maropu, password=, driver=org.postgresql.Driver})[col0#0,col1#1] PushedFilters: [EqualTo(col0,xxx)]
```

Author: Takeshi YAMAMURO <linguin.m.s@gmail.com>

Closes #10427 from maropu/RemoveFilterInJdbcScan.

zsxwing committed:
[SPARK-13149][SQL] Add FileStreamSource

`FileStreamSource` is an implementation of `org.apache.spark.sql.execution.streaming.Source`. It takes advantage of the existing `HadoopFsRelationProvider` to support various file formats. It remembers files in each batch and stores it into the metadata files so as to recover them when restarting. The metadata files are stored in the file system. There will be a further PR to clean up the metadata files periodically.

This is based on the initial work from marmbrus.

Author: Shixiong Zhu <shixiong@databricks.com>

Closes #11034 from zsxwing/stream-df-file-source.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

SparkQA commented:
**[Test build #51024 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51024/consoleFull)** for PR 11140 at commit [`a400264`](https://github.com/apache/spark/commit/a40026432c92462f5e50916bd64160829114e48d).

SparkQA commented:
**[Test build #51024 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51024/consoleFull)** for PR 11140 at commit [`a400264`](https://github.com/apache/spark/commit/a40026432c92462f5e50916bd64160829114e48d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51024/
Test FAILed.

zsxwing commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -889,30 +910,31 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
-          logWarning(s"Error deleting ${file.getPath}")
+  /** Create a seperate daemon thread to remove phantomreferences from queue and invoke cleanup */
+      
+ val referenceThread = new Thread {
```
nit: set a thread name to help debug

gliptak committed:
[SPARK-11565] Replace deprecated DigestUtils.shaHex call

Author: Gábor Lipták <gliptak@gmail.com>

Closes #9532 from gliptak/SPARK-11565.

tritab committed:
[SPARK-11518][DEPLOY, WINDOWS] Handle spaces in Windows command scripts

Author: Jon Maurer <tritab@gmail.com>
Author: Jonathan Maurer <jmaurer@Jonathans-MacBook-Pro.local>

Closes #10789 from tritab/cmd_updates.

tedyu committed:
[SPARK-13203] Add scalastyle rule banning use of mutable.SynchronizedBuffer

andrewor14
Please take a look

Author: tedyu <yuzhihong@gmail.com>

Closes #11134 from tedyu/master.

srowen committed:
[SPARK-9307][CORE][SPARK] Logging: Make it either stable or private

Make Logging private[spark]. Pretty much all there is to it.

Author: Sean Owen <sowen@cloudera.com>

Closes #11103 from srowen/SPARK-9307.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

SparkQA commented:
**[Test build #51038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51038/consoleFull)** for PR 11140 at commit [`0af18af`](https://github.com/apache/spark/commit/0af18af63e0359c8971bf1f8eeac3f0e8c137070).

SparkQA commented:
**[Test build #51038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51038/consoleFull)** for PR 11140 at commit [`0af18af`](https://github.com/apache/spark/commit/0af18af63e0359c8971bf1f8eeac3f0e8c137070).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51038/
Test FAILed.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

SparkQA commented:
**[Test build #51041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51041/consoleFull)** for PR 11140 at commit [`7068425`](https://github.com/apache/spark/commit/7068425c27e086040dd73e73ec079e8da749754d).

SparkQA commented:
**[Test build #51041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51041/consoleFull)** for PR 11140 at commit [`7068425`](https://github.com/apache/spark/commit/7068425c27e086040dd73e73ec079e8da749754d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51041/
Test FAILed.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

Commit:
[SPARK-5095][MESOS] Support launching multiple mesos executors in coarse grained mesos mode.

This is the next iteration of tnachen's previous PR: https://github.com/apache/spark/pull/4027

In that PR, we resolved with andrewor14 and pwendell to implement the Mesos scheduler's support of `spark.executor.cores` to be consistent with YARN and Standalone.  This PR implements that resolution.

This PR implements two high-level features.  These two features are co-dependent, so they're implemented both here:
- Mesos support for spark.executor.cores
- Multiple executors per slave

We at Mesosphere have been working with Typesafe on a Spark/Mesos integration test suite: https://github.com/typesafehub/mesos-spark-integration-tests, which passes for this PR.

The contribution is my original work and I license the work to the project under the project's open source license.

Author: Michael Gummelt <mgummelt@mesosphere.io>

Closes #10993 from mgummelt/executor_sizing.

SparkQA commented:
**[Test build #51044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51044/consoleFull)** for PR 11140 at commit [`1b3f732`](https://github.com/apache/spark/commit/1b3f732df753b0569137f34b6a4c4c8d5773c440).

JoshRosen committed:
[SPARK-13254][SQL] Fix planning of TakeOrderedAndProject operator

The patch for SPARK-8964 ("use Exchange to perform shuffle in Limit" / #7334) inadvertently broke the planning of the TakeOrderedAndProject operator: because ReturnAnswer was the new root of the query plan, the TakeOrderedAndProject rule was unable to match before BasicOperators.

This patch fixes this by moving the `TakeOrderedAndCollect` and `CollectLimit` rules into the same strategy.

In addition, I made changes to the TakeOrderedAndProject operator in order to make its `doExecute()` method lazy and added a new TakeOrderedAndProjectSuite which tests the new code path.

/cc davies and marmbrus for review.

Author: Josh Rosen <joshrosen@databricks.com>

Closes #11145 from JoshRosen/take-ordered-and-project-fix.

ajbozarth committed:
[SPARK-13163][WEB UI] Column width on new History Server DataTables not getting set correctly

The column width for the new DataTables now adjusts for the current page rather than being hard-coded for the entire table's data.

Author: Alex Bozarth <ajbozart@us.ibm.com>

Closes #11057 from ajbozarth/spark13163.

Commit:
[SPARK-13126] fix the right margin of history page.

The right margin of the history page is little bit off. A simple fix for that issue.

Author: zhuol <zhuol@yahoo-inc.com>

Closes #11029 from zhuoliu/13126.

JoshRosen committed:
[HOTFIX] Fix Scala 2.10 build break in TakeOrderedAndProjectSuite.

SparkQA commented:
**[Test build #51044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51044/consoleFull)** for PR 11140 at commit [`1b3f732`](https://github.com/apache/spark/commit/1b3f732df753b0569137f34b6a4c4c8d5773c440).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51044/
Test FAILed.

maropu committed:
[SPARK-13057][SQL] Add benchmark codes and the performance results for implemented compression schemes for InMemoryRelation

This pr adds benchmark codes for in-memory cache compression to make future developments and discussions more smooth.

Author: Takeshi YAMAMURO <linguin.m.s@gmail.com>

Closes #10965 from maropu/ImproveColumnarCache.

srowen committed:
[SPARK-12414][CORE] Remove closure serializer

Remove spark.closure.serializer option and use JavaSerializer always

CC andrewor14 rxin I see there's a discussion in the JIRA but just thought I'd offer this for a look at what the change would be.

Author: Sean Owen <sowen@cloudera.com>

Closes #11150 from srowen/SPARK-12414.

rxin commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -18,12 +18,17 @@
+import scala.collection.mutable.ListBuffer
 import scala.language.existentials
+import scala.language.postfixOps
+import scala.ref
```
what's this import for?

rxin commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -874,10 +890,14 @@ private class PythonAccumulatorParam(@transient private val serverHost: String,
+  val queue = new ReferenceQueue[File]()
+  val phantomReferences = new ListBuffer[FilePhantomReference]()
+  val threadName = "WeakReference"
+/**
```
can you fix the indentation here, and add proper blank lines (line before the doc, not after)

rxin commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -889,30 +909,32 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
    * Write data into disk, using randomly generated name.
    */
-  private def readObject(in: ObjectInputStream): Unit = Utils.tryOrIOException {
+ private def readObject(in: ObjectInputStream): Unit = Utils.tryOrIOException {
```
fix indentation

rxin commented:
@GayathriMurali thanks for doing this, but can you please fix all the style violations you have? The changes should indent the code blocks consistently.

tdas committed:
[SPARK-13146][SQL] Management API for continuous queries

### Management API for Continuous Queries

**API for getting status of each query**
- Whether active or not
- Unique name of each query
- Status of the sources and sinks
- Exceptions

**API for managing each query**
- Immediately stop an active query
- Waiting for a query to be terminated, correctly or with error

**API for managing multiple queries**
- Listing all active queries
- Getting an active query by name
- Waiting for any one of the active queries to be terminated

**API for listening to query life cycle events**
- ContinuousQueryListener API for query start, progress and termination events.

Author: Tathagata Das <tathagata.das1565@gmail.com>

Closes #11030 from tdas/streaming-df-management-api.

raelawang committed:
[SPARK-13274] Fix Aggregator Links on GroupedDataset Scala API

Update Aggregator links to point to #org.apache.spark.sql.expressions.Aggregator

Author: raela <raela@databricks.com>

Closes #11158 from raelawang/master.

gatorsmile committed:
[SPARK-12725][SQL] Resolving Name Conflicts in SQL Generation and Name Ambiguity Caused by Internally Generated Expressions

Some analysis rules generate aliases or auxiliary attribute references with the same name but different expression IDs. For example, `ResolveAggregateFunctions` introduces `havingCondition` and `aggOrder`, and `DistinctAggregationRewriter` introduces `gid`.

This is OK for normal query execution since these attribute references get expression IDs. However, it's troublesome when converting resolved query plans back to SQL query strings since expression IDs are erased.

Here's an example Spark 1.6.0 snippet for illustration:
```scala
sqlContext.range(10).select('id as 'a, 'id as 'b).registerTempTable("t")
sqlContext.sql("SELECT SUM(a) FROM t GROUP BY a, b ORDER BY COUNT(a), COUNT(b)").explain(true)
```
The above code produces the following resolved plan:
```
== Analyzed Logical Plan ==
_c0: bigint
Project [_c0#101L]
+- Sort [aggOrder#102L ASC,aggOrder#103L ASC], true
   +- Aggregate [a#47L,b#48L], [(sum(a#47L),mode=Complete,isDistinct=false) AS _c0#101L,(count(a#47L),mode=Complete,isDistinct=false) AS aggOrder#102L,(count(b#48L),mode=Complete,isDistinct=false) AS aggOrder#103L]
      +- Subquery t
         +- Project [id#46L AS a#47L,id#46L AS b#48L]
            +- LogicalRDD [id#46L], MapPartitionsRDD[44] at range at <console>:26
```
Here we can see that both aggregate expressions in `ORDER BY` are extracted into an `Aggregate` operator, and both of them are named `aggOrder` with different expression IDs.

The solution is to automatically add the expression IDs into the attribute name for the Alias and AttributeReferences that are generated by Analyzer in SQL Generation.

In this PR, it also resolves another issue. Users could use the same name as the internally generated names. The duplicate names should not cause name ambiguity. When resolving the column, Catalyst should not pick the column that is internally generated.

Could you review the solution? marmbrus liancheng

I did not set the newly added flag for all the alias and attribute reference generated by Analyzers. Please let me know if I should do it? Thank you!

Author: gatorsmile <gatorsmile@gmail.com>

Closes #11050 from gatorsmile/namingConflicts.

gatorsmile committed:
[SPARK-13205][SQL] SQL Generation Support for Self Join

This PR addresses two issues:
  - Self join does not work in SQL Generation
  - When creating new instances for `LogicalRelation`, `metastoreTableIdentifier` is lost.

liancheng Could you please review the code changes? Thank you!

Author: gatorsmile <gatorsmile@gmail.com>

Closes #11084 from gatorsmile/selfJoinInSQLGen.

Commit:
[SPARK-12706] [SQL] grouping() and grouping_id()

Grouping() returns a column is aggregated or not, grouping_id() returns the aggregation levels.

grouping()/grouping_id() could be used with window function, but does not work in having/sort clause, will be fixed by another PR.

The GROUPING__ID/grouping_id() in Hive is wrong (according to docs), we also did it wrongly, this PR change that to match the behavior in most databases (also the docs of Hive).

Author: Davies Liu <davies@databricks.com>

Closes #10677 from davies/grouping.

Commit:
[SPARK-13234] [SQL] remove duplicated SQL metrics

For lots of SQL operators, we have metrics for both of input and output, the number of input rows should be exactly the number of output rows of child, we could only have metrics for output rows.

After we improved the performance using whole stage codegen, the overhead of SQL metrics are not trivial anymore, we should avoid that if it's not necessary.

This PR remove all the SQL metrics for number of input rows, add SQL metric of number of output rows for all LeafNode. All remove the SQL metrics from those operators that have the same number of rows from input and output (for example, Projection, we may don't need that).

The new SQL UI will looks like:

![metrics](https://cloud.githubusercontent.com/assets/40902/12965227/63614e5e-d009-11e5-88b3-84fea04f9c20.png)

Author: Davies Liu <davies@databricks.com>

Closes #11163 from davies/remove_metrics.

hvanhovell committed:
[SPARK-13276] Catch bad characters at the end of a Table Identifier/Expression string

The parser currently parses the following strings without a hitch:
* Table Identifier:
  * `a.b.c` should fail, but results in the following table identifier `a.b`
  * `table!#` should fail, but results in the following table identifier `table`
* Expression
  * `1+2 r+e` should fail, but results in the following expression `1 + 2`

This PR fixes this by adding terminated rules for both expression parsing and table identifier parsing.

cc cloud-fan (we discussed this in https://github.com/apache/spark/pull/10649) jayadevanmurali (this causes your PR https://github.com/apache/spark/pull/11051 to fail)

Author: Herman van Hovell <hvanhovell@questtec.nl>

Closes #11159 from hvanhovell/SPARK-13276.

gatorsmile committed:
[SPARK-13235][SQL] Removed an Extra Distinct from the Plan when Using Union in SQL

Currently, the parser added two `Distinct` operators in the plan if we are using `Union` or `Union Distinct` in the SQL. This PR is to remove the extra `Distinct` from the plan.

For example, before the fix, the following query has a plan with two `Distinct`
```scala
sql("select * from t0 union select * from t0").explain(true)
```
```
== Parsed Logical Plan ==
'Project [unresolvedalias(*,None)]
+- 'Subquery u_2
   +- 'Distinct
      +- 'Project [unresolvedalias(*,None)]
         +- 'Subquery u_1
            +- 'Distinct
               +- 'Union
                  :- 'Project [unresolvedalias(*,None)]
                  :  +- 'UnresolvedRelation `t0`, None
                  +- 'Project [unresolvedalias(*,None)]
                     +- 'UnresolvedRelation `t0`, None

== Analyzed Logical Plan ==
id: bigint
Project [id#16L]
+- Subquery u_2
   +- Distinct
      +- Project [id#16L]
         +- Subquery u_1
            +- Distinct
               +- Union
                  :- Project [id#16L]
                  :  +- Subquery t0
                  :     +- Relation[id#16L] ParquetRelation
                  +- Project [id#16L]
                     +- Subquery t0
                        +- Relation[id#16L] ParquetRelation

== Optimized Logical Plan ==
Aggregate [id#16L], [id#16L]
+- Aggregate [id#16L], [id#16L]
   +- Union
      :- Project [id#16L]
      :  +- Relation[id#16L] ParquetRelation
      +- Project [id#16L]
         +- Relation[id#16L] ParquetRelation
```
After the fix, the plan is changed without the extra `Distinct` as follows:
```
== Parsed Logical Plan ==
'Project [unresolvedalias(*,None)]
+- 'Subquery u_1
   +- 'Distinct
      +- 'Union
         :- 'Project [unresolvedalias(*,None)]
         :  +- 'UnresolvedRelation `t0`, None
         +- 'Project [unresolvedalias(*,None)]
           +- 'UnresolvedRelation `t0`, None

== Analyzed Logical Plan ==
id: bigint
Project [id#17L]
+- Subquery u_1
   +- Distinct
      +- Union
        :- Project [id#16L]
        :  +- Subquery t0
        :     +- Relation[id#16L] ParquetRelation
        +- Project [id#16L]
          +- Subquery t0
          +- Relation[id#16L] ParquetRelation

== Optimized Logical Plan ==
Aggregate [id#17L], [id#17L]
+- Union
  :- Project [id#16L]
  :  +- Relation[id#16L] ParquetRelation
  +- Project [id#16L]
    +- Relation[id#16L] ParquetRelation
```

Author: gatorsmile <gatorsmile@gmail.com>

Closes #11120 from gatorsmile/unionDistinct.

nongli committed:
[SPARK-13270][SQL] Remove extra new lines in whole stage codegen and include pipeline plan in comments.

Author: Nong Li <nong@databricks.com>

Closes #11155 from nongli/spark-13270.

sasakitoa committed:
[SPARK-13264][DOC] Removed multi-byte characters in spark-env.sh.template

In spark-env.sh.template, there are multi-byte characters, this PR will remove it.

Author: Sasaki Toru <sasakitoa@nttdata.co.jp>

Closes #11149 from sasakitoa/remove_multibyte_in_sparkenv.

jyssky committed:
[SPARK-13074][CORE] Add JavaSparkContext. getPersistentRDDs method

The "getPersistentRDDs()" is a useful API of SparkContext to get cached RDDs. However, the JavaSparkContext does not have this API.

Add a simple getPersistentRDDs() to get java.util.Map<Integer, JavaRDD> for Java users.

Author: Junyang <fly.shenjy@gmail.com>

Closes #10978 from flyjy/master.

ajbozarth committed:
[SPARK-13124][WEB UI] Fixed CSS and JS issues caused by addition of JQuery DataTables

Made sure the old tables continue to use the old css and the new DataTables use the new css. Also fixed it so the Safari Web Inspector doesn't throw errors when on the new DataTables pages.

Author: Alex Bozarth <ajbozart@us.ibm.com>

Closes #11038 from ajbozarth/spark13124.

tdas committed:
[STREAMING][TEST] Fix flaky streaming.FailureSuite

Under some corner cases, the test suite failed to shutdown the SparkContext causing cascaded failures. This fix does two things
- Makes sure no SparkContext is active after every test
- Makes sure StreamingContext is always shutdown (prevents leaking of StreamingContexts as well, just in case)

Author: Tathagata Das <tathagata.das1565@gmail.com>

Closes #11166 from tdas/fix-failuresuite.

viirya committed:
[SPARK-13277][SQL] ANTLR ignores other rule using the USING keyword

JIRA: https://issues.apache.org/jira/browse/SPARK-13277

There is an ANTLR warning during compilation:

    warning(200): org/apache/spark/sql/catalyst/parser/SparkSqlParser.g:938:7:
    Decision can match input such as "KW_USING Identifier" using multiple alternatives: 2, 3

    As a result, alternative(s) 3 were disabled for that input

This patch is to fix it.

Author: Liang-Chi Hsieh <viirya@gmail.com>

Closes #11168 from viirya/fix-parser-using.

jayadevanmurali committed:
[SPARK-12982][SQL] Add table name validation in temp table registration

Add the table name validation at the temp table creation

Author: jayadevanmurali <jayadevan.m@tcs.com>

Closes #11051 from jayadevanmurali/branch-0.2-SPARK-12982.

Commit:
[SPARK-13279] Remove O(n^2) operation from scheduler.

This commit removes an unnecessary duplicate check in addPendingTask that meant
that scheduling a task set took time proportional to (# tasks)^2.

Author: Sital Kedia <skedia@fb.com>

Closes #11167 from sitalkedia/fix_stuck_driver and squashes the following commits:

3fe1af8 [Sital Kedia] [SPARK-13279] Remove unnecessary duplicate check in addPendingTask function

rxin committed:
Revert "[SPARK-13279] Remove O(n^2) operation from scheduler."

This reverts commit 50fa6fd1b365d5db7e2b2c59624a365cef0d1696.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

SparkQA commented:
**[Test build #51134 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51134/consoleFull)** for PR 11140 at commit [`837252a`](https://github.com/apache/spark/commit/837252a74ec87e8f1ac07e80406bf0410c9088d7).

yu-iskw committed:
[SPARK-13265][ML] Refactoring of basic ML import/export for other file system besides HDFS

jkbradley I tried to improve the function to export a model. When I tried to export a model to S3 under Spark 1.6, we couldn't do that. So, it should offer S3 besides HDFS. Can you review it when you have time? Thanks!

Author: Yu ISHIKAWA <yuu.ishikawa@gmail.com>

Closes #11151 from yu-iskw/SPARK-13265.

yu-iskw committed:
[SPARK-11515][ML] QuantileDiscretizer should take random seed

cc jkbradley

Author: Yu ISHIKAWA <yuu.ishikawa@gmail.com>

Closes #9535 from yu-iskw/SPARK-11515.

vectorijk committed:
[SPARK-13037][ML][PYSPARK] PySpark ml.recommendation support export/import

PySpark ml.recommendation support export/import.

Author: Kai Jiang <jiangkai@gmail.com>

Closes #11044 from vectorijk/spark-13037.

yanboliang committed:
[MINOR][ML][PYSPARK] Cleanup test cases of clustering.py

Test cases should be removed from annotation of ```setXXX``` function, otherwise it will be parts of [Python API docs](https://spark.apache.org/docs/latest/api/python/pyspark.ml.html#pyspark.ml.clustering.KMeans.setInitMode).
cc mengxr jkbradley

Author: Yanbo Liang <ybliang8@gmail.com>

Closes #10975 from yanboliang/clustering-cleanup.

yanboliang committed:
[SPARK-13035][ML][PYSPARK] PySpark ml.clustering support export/import

PySpark ml.clustering support export/import.

Author: Yanbo Liang <ybliang8@gmail.com>

Closes #10999 from yanboliang/spark-13035.

sethah committed:
[SPARK-13047][PYSPARK][ML] Pyspark Params.hasParam should not throw an error

Pyspark Params class has a method `hasParam(paramName)` which returns `True` if the class has a parameter by that name, but throws an `AttributeError` otherwise. There is not currently a way of getting a Boolean to indicate if a class has a parameter. With Spark 2.0 we could modify the existing behavior of `hasParam` or add an additional method with this functionality.

In Python:
```python
from pyspark.ml.classification import NaiveBayes
nb = NaiveBayes()
print nb.hasParam("smoothing")
print nb.hasParam("notAParam")
```
produces:
> True
> AttributeError: 'NaiveBayes' object has no attribute 'notAParam'

However, in Scala:
```scala
import org.apache.spark.ml.classification.NaiveBayes
val nb  = new NaiveBayes()
nb.hasParam("smoothing")
nb.hasParam("notAParam")
```
produces:
> true
> false

cc holdenk

Author: sethah <seth.hendrickson16@gmail.com>

Closes #10962 from sethah/SPARK-13047.

SparkQA commented:
**[Test build #51134 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51134/consoleFull)** for PR 11140 at commit [`837252a`](https://github.com/apache/spark/commit/837252a74ec87e8f1ac07e80406bf0410c9088d7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51134/
Test FAILed.

GayathriMurali commented:
![image](https://cloud.githubusercontent.com/assets/7002441/12996389/39b20914-d0e4-11e5-8be5-5d6da97baf6e.png)

Test build failed spark unit tests. But I am not able to see any failures in the build results.

sloth2012 committed:
[SPARK-12765][ML][COUNTVECTORIZER] fix CountVectorizer.transform's lost transformSchema

https://issues.apache.org/jira/browse/SPARK-12765

Author: Liu Xiang <lxmtlab@gmail.com>

Closes #10720 from sloth2012/sloth.

Commit:
[SPARK-12915][SQL] add SQL metrics of numOutputRows for whole stage codegen

This PR add SQL metrics (numOutputRows) for generated operators (same as non-generated), the cost is about 0.2 nano seconds per row.

<img width="806" alt="gen metrics" src="https://cloud.githubusercontent.com/assets/40902/12994694/47f5881e-d0d7-11e5-9d47-78229f559ab0.png">

Author: Davies Liu <davies@databricks.com>

Closes #11170 from davies/gen_metric.

hvanhovell committed:
[SPARK-13277][BUILD] Follow-up ANTLR warnings are treated as build errors

It is possible to create faulty but legal ANTLR grammars. ANTLR will produce warnings but also a valid compileable parser. This PR makes sure we treat such warnings as build errors.

cc rxin / viirya

Author: Herman van Hovell <hvanhovell@questtec.nl>

Closes #11174 from hvanhovell/ANTLR-warnings-as-errors.

Earthson committed:
[SPARK-12746][ML] ArrayType(_, true) should also accept ArrayType(_, false)

https://issues.apache.org/jira/browse/SPARK-12746

Author: Earthson Lu <Earthson.Lu@gmail.com>

Closes #10697 from Earthson/SPARK-12746.

Wenpei committed:
[SPARK-13153][PYSPARK] ML persistence failed when handle no default value parameter

Fix this defect by check default value exist or not.

yanboliang Please help to review.

Author: Tommy YU <tummyyu@163.com>

Closes #11043 from Wenpei/spark-13153-handle-param-withnodefaultvalue.

steveloughran committed:
[SPARK-7889][WEBUI] HistoryServer updates UI for incomplete apps

When the HistoryServer is showing an incomplete app, it needs to check if there is a newer version of the app available.  It does this by checking if a version of the app has been loaded with a larger *filesize*.  If so, it detaches the current UI, attaches the new one, and redirects back to the same URL to show the new UI.

https://issues.apache.org/jira/browse/SPARK-7889

Author: Steve Loughran <stevel@hortonworks.com>
Author: Imran Rashid <irashid@cloudera.com>

Closes #11118 from squito/SPARK-7889-alternate.

zsxwing commented:
retest this please

Commit:
[SPARK-6166] Limit number of in flight outbound requests

This JIRA is related to
https://github.com/apache/spark/pull/5852
Had to do some minor rework and test to make sure it
works with current version of spark.

Author: Sanket <schintap@untilservice-lm>

Closes #10838 from redsanket/limit-outbound-connections.

SparkQA commented:
**[Test build #51175 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51175/consoleFull)** for PR 11140 at commit [`837252a`](https://github.com/apache/spark/commit/837252a74ec87e8f1ac07e80406bf0410c9088d7).

SparkQA commented:
**[Test build #51175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51175/consoleFull)** for PR 11140 at commit [`837252a`](https://github.com/apache/spark/commit/837252a74ec87e8f1ac07e80406bf0410c9088d7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51175/
Test FAILed.

yanboliang committed:
[SPARK-12974][ML][PYSPARK] Add Python API for spark.ml bisecting k-means

Add Python API for spark.ml bisecting k-means.

Author: Yanbo Liang <ybliang8@gmail.com>

Closes #10889 from yanboliang/spark-12974.

holdenk committed:
[SPARK-13154][PYTHON] Add linting for pydocs

We should have lint rules using sphinx to automatically catch the pydoc issues that are sometimes introduced.

Right now ./dev/lint-python will skip building the docs if sphinx isn't present - but it might make sense to fail hard - just a matter of if we want to insist all PySpark developers have sphinx present.

Author: Holden Karau <holden@us.ibm.com>

Closes #11109 from holdenk/SPARK-13154-add-pydoc-lint-for-docs.

Commit:
[SPARK-12705] [SQL] push missing attributes for Sort

The current implementation of ResolveSortReferences can only push one missing attributes into it's child, it failed to analyze TPCDS Q98, because of there are two missing attributes in that (one from Window, another from Aggregate).

Author: Davies Liu <davies@databricks.com>

Closes #11153 from davies/resolve_sort.

rxin committed:
[SPARK-13282][SQL] LogicalPlan toSql should just return a String

Previously we were using Option[String] and None to indicate the case when Spark fails to generate SQL. It is easier to just use exceptions to propagate error cases, rather than having for comprehension everywhere. I also introduced a "build" function that simplifies string concatenation (i.e. no need to reason about whether we have an extra space or not).

Author: Reynold Xin <rxin@databricks.com>

Closes #11171 from rxin/SPARK-13282.

HyukjinKwon committed:
[SPARK-13260][SQL] count(*) does not work with CSV data source

https://issues.apache.org/jira/browse/SPARK-13260
This is a quicky fix for `count(*)`.

When the `requiredColumns` is empty, currently it returns `sqlContext.sparkContext.emptyRDD[Row]` which does not have the count.

Just like JSON datasource, this PR lets the CSV datasource count the rows but do not parse each set of tokens.

Author: hyukjinkwon <gurwls223@gmail.com>

Closes #11169 from HyukjinKwon/SPARK-13260.

yanboliang committed:
[SPARK-12962] [SQL] [PySpark] PySpark support covar_samp and covar_pop

PySpark support ```covar_samp``` and ```covar_pop```.

cc rxin davies marmbrus

Author: Yanbo Liang <ybliang8@gmail.com>

Closes #10876 from yanboliang/spark-12962.

vijaykiran committed:
[SPARK-12630][PYSPARK] [DOC] PySpark classification parameter desc to consistent format

Part of task for [SPARK-11219](https://issues.apache.org/jira/browse/SPARK-11219) to make PySpark MLlib parameter description formatting consistent. This is for the classification module.

Author: vijaykiran <mail@vijaykiran.com>
Author: Bryan Cutler <cutlerb@gmail.com>

Closes #11183 from BryanCutler/pyspark-consistent-param-classification-SPARK-12630.

Commit:
[SPARK-5095] Fix style in mesos coarse grained scheduler code

andrewor14 This addressed your style comments from #10993

Author: Michael Gummelt <mgummelt@mesosphere.io>

Closes #11187 from mgummelt/fix_mesos_style.

Commit:
[SPARK-5095] remove flaky test

Overrode the start() method, which was previously starting a thread causing a race condition. I believe this should fix the flaky test.

Author: Michael Gummelt <mgummelt@mesosphere.io>

Closes #11164 from mgummelt/fix_mesos_tests.

zsxwing commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -892,27 +911,30 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
       Utils.copyStream(in, out)
     } {
-      out.close()
+       in.close();
```
Why change this line? I think this line makes PythonBroadcastSuite fail.

zsxwing commented on core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala:
```diff
@@ -892,27 +911,30 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
-        if (!file.delete()) {
-          logWarning(s"Error deleting ${file.getPath}")
-        }
+    val referenceThread = new Thread(threadName){
```
This will create a thread for each `PythonBroadcast`. We should avoid that.

Maybe use a global thread to do that? However, this will leak a thread even if it's not used any more (e.g., after stopping SparkContext).

Maybe `finalize` is the best option.

GayathriMurali commented:
@zsxwing I dont have a constructor to do object initialization for the FilePhantomReferences class. I am guessing that could also be causing python broadcast to fail with null pointer exception. I am going to fix that. I will also remove the change on 920(out.close), I tried removing that line earlier to test something and added the  wrong one.

Commit:
[SPARK-13293][SQL] generate Expand

Expand suffer from create the UnsafeRow from same input multiple times, with codegen, it only need to copy some of the columns.

After this, we can see 3X improvements (from 43 seconds to 13 seconds) on a TPCDS query (Q67) that have eight columns in Rollup.

Ideally, we could mask some of the columns based on bitmask, I'd leave that in the future, because currently Aggregation (50 ns) is much slower than that just copy the variables (1-2 ns).

Author: Davies Liu <davies@databricks.com>

Closes #11177 from davies/gen_expand.

Commit:
[SPARK-13142][WEB UI] Problem accessing Web UI /logPage/ on Microsoft Windows

Due to being on a Windows platform I have been unable to run the tests as described in the "Contributing to Spark" instructions. As the change is only to two lines of code in the Web UI, which I have manually built and tested, I am submitting this pull request anyway. I hope this is OK.

Is it worth considering also including this fix in any future 1.5.x releases (if any)?

I confirm this is my own original work and license it to the Spark project under its open source license.

Author: markpavey <mark.pavey@thefilter.com>

Closes #11135 from markpavey/JIRA_SPARK-13142_WindowsWebUILogFix.

viirya committed:
[SPARK-12363][MLLIB] Remove setRun and fix PowerIterationClustering failed test

JIRA: https://issues.apache.org/jira/browse/SPARK-12363

This issue is pointed by yanboliang. When `setRuns` is removed from PowerIterationClustering, one of the tests will be failed. I found that some `dstAttr`s of the normalized graph are not correct values but 0.0. By setting `TripletFields.All` in `mapTriplets` it can work.

Author: Liang-Chi Hsieh <viirya@gmail.com>
Author: Xiangrui Meng <meng@databricks.com>

Closes #10539 from viirya/fix-poweriter.

rxin committed:
Closes #11185

srowen committed:
[SPARK-13172][CORE][SQL] Stop using RichException.getStackTrace it is deprecated

Replace `getStackTraceString` with `Utils.exceptionString`

Author: Sean Owen <sowen@cloudera.com>

Closes #11182 from srowen/SPARK-13172.

rxin committed:
[SPARK-13296][SQL] Move UserDefinedFunction into sql.expressions.

This pull request has the following changes:

1. Moved UserDefinedFunction into expressions package. This is more consistent with how we structure the packages for window functions and UDAFs.

2. Moved UserDefinedPythonFunction into execution.python package, so we don't have a random private class in the top level sql package.

3. Move everything in execution/python.scala into the newly created execution.python package.

Most of the diffs are just straight copy-paste.

Author: Reynold Xin <rxin@databricks.com>

Closes #11181 from rxin/SPARK-13296.

amitdev committed:
[SPARK-13300][DOCUMENTATION] Added pygments.rb dependancy

Looks like pygments.rb gem is also required for jekyll build to work. At least on Ubuntu/RHEL I could not do build without this dependency. So added this to steps.

Author: Amit Dev <amitdev@gmail.com>

Closes #11180 from amitdev/master.

cl4es committed:
[SPARK-13278][CORE] Launcher fails to start with JDK 9 EA

See http://openjdk.java.net/jeps/223 for more information about the JDK 9 version string scheme.

Author: Claes Redestad <claes.redestad@gmail.com>

Closes #11160 from cl4es/master.

carsonwang committed:
[SPARK-13185][SQL] Reuse Calendar object in DateTimeUtils.StringToDate method to improve performance

The java `Calendar` object is expensive to create. I have a sub query like this `SELECT a, b, c FROM table UV WHERE (datediff(UV.visitDate, '1997-01-01')>=0 AND datediff(UV.visitDate, '2015-01-01')<=0))`

The table stores `visitDate` as String type and has 3 billion records. A `Calendar` object is created every time `DateTimeUtils.stringToDate` is called. By reusing the `Calendar` object, I saw about 20 seconds performance improvement for this stage.

Author: Carson Wang <carson.wang@intel.com>

Closes #11090 from carsonwang/SPARK-13185.

JoshRosen committed:
[SPARK-12503][SPARK-12505] Limit pushdown in UNION ALL and OUTER JOIN

This patch adds a new optimizer rule for performing limit pushdown. Limits will now be pushed down in two cases:

- If a limit is on top of a `UNION ALL` operator, then a partition-local limit operator will be pushed to each of the union operator's children.
- If a limit is on top of an `OUTER JOIN` then a partition-local limit will be pushed to one side of the join. For `LEFT OUTER` and `RIGHT OUTER` joins, the limit will be pushed to the left and right side, respectively. For `FULL OUTER` join, we will only push limits when at most one of the inputs is already limited: if one input is limited we will push a smaller limit on top of it and if neither input is limited then we will limit the input which is estimated to be larger.

These optimizations were proposed previously by gatorsmile in #10451 and #10454, but those earlier PRs were closed and deferred for later because at that time Spark's physical `Limit` operator would trigger a full shuffle to perform global limits so there was a chance that pushdowns could actually harm performance by causing additional shuffles/stages. In #7334, we split the `Limit` operator into separate `LocalLimit` and `GlobalLimit` operators, so we can now push down only local limits (which don't require extra shuffles). This patch is based on both of gatorsmile's patches, with changes and simplifications due to partition-local-limiting.

When we push down the limit, we still keep the original limit in place, so we need a mechanism to ensure that the optimizer rule doesn't keep pattern-matching once the limit has been pushed down. In order to handle this, this patch adds a `maxRows` method to `SparkPlan` which returns the maximum number of rows that the plan can compute, then defines the pushdown rules to only push limits to children if the children's maxRows are greater than the limit's maxRows. This idea is carried over from #10451; see that patch for additional discussion.

Author: Josh Rosen <joshrosen@databricks.com>

Closes #11121 from JoshRosen/limit-pushdown-2.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

SparkQA commented:
**[Test build #51293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51293/consoleFull)** for PR 11140 at commit [`7f9dffc`](https://github.com/apache/spark/commit/7f9dffc0e98fe12a7c4e6c2373048da147df01ee).

SparkQA commented:
**[Test build #51293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51293/consoleFull)** for PR 11140 at commit [`7f9dffc`](https://github.com/apache/spark/commit/7f9dffc0e98fe12a7c4e6c2373048da147df01ee).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51293/
Test FAILed.

GayathriMurali committed:
SPARK-12729 PhantomReferences to replace Finalize()- Fixed styling issues

SparkQA commented:
**[Test build #51294 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51294/consoleFull)** for PR 11140 at commit [`44b16af`](https://github.com/apache/spark/commit/44b16af2b803a9d9eea2dcbfcd631b5899f9314d).

SparkQA commented:
**[Test build #51294 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51294/consoleFull)** for PR 11140 at commit [`44b16af`](https://github.com/apache/spark/commit/44b16af2b803a9d9eea2dcbfcd631b5899f9314d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51294/
Test FAILed.

maropu committed:
[SPARK-12995][GRAPHX] Remove deprecate APIs from Pregel

Author: Takeshi YAMAMURO <linguin.m.s@gmail.com>

Closes #10918 from maropu/RemoveDeprecateInPregel.

JeremyNixon committed:
[SPARK-13312][MLLIB] Update java train-validation-split example in ml-guide

Response to JIRA https://issues.apache.org/jira/browse/SPARK-13312.

This contribution is my original work and I license the work to this project.

Author: JeremyNixon <jnixon2@gmail.com>

Closes #11199 from JeremyNixon/update_train_val_split_example.

seddonm1 committed:
[SPARK-13097][ML] Binarizer allowing Double AND Vector input types

This enhancement extends the existing SparkML Binarizer [SPARK-5891] to allow Vector in addition to the existing Double input column type.

A use case for this enhancement is for when a user wants to Binarize many similar feature columns at once using the same threshold value (for example a binary threshold applied to many pixels in an image).

This contribution is my original work and I license the work to the project under the project's open source license.

viirya mengxr

Author: seddonm1 <seddonm1@gmail.com>

Closes #10976 from seddonm1/master.

keypointt committed:
[SPARK-13018][DOCS] Replace example code in mllib-pmml-model-export.md using include_example

Replace example code in mllib-pmml-model-export.md using include_example
https://issues.apache.org/jira/browse/SPARK-13018

The example code in the user guide is embedded in the markdown and hence it is not easy to test. It would be nice to automatically test them. This JIRA is to discuss options to automate example code testing and see what we can do in Spark 1.6.

Goal is to move actual example code to spark/examples and test compilation in Jenkins builds. Then in the markdown, we can reference part of the code to show in the user guide. This requires adding a Jekyll tag that is similar to https://github.com/jekyll/jekyll/blob/master/lib/jekyll/tags/include.rb, e.g., called include_example.
`{% include_example scala/org/apache/spark/examples/mllib/PMMLModelExportExample.scala %}`
Jekyll will find `examples/src/main/scala/org/apache/spark/examples/mllib/PMMLModelExportExample.scala` and pick code blocks marked "example" and replace code block in
`{% highlight %}`
 in the markdown.

See more sub-tasks in parent ticket: https://issues.apache.org/jira/browse/SPARK-11337

Author: Xin Ren <iamshrek@126.com>

Closes #11126 from keypointt/SPARK-13018.

JoshRosen commented:
No comment on the contents of this PR (since I haven't looked at them), but I did want to note that I think that the pull request description is a little thin here. Could you add a concise summary of the changes here, their impact on the code, and motivation for why we're doing this? This helps reviewers / readers know what to focus on and also helps future readers by allowing them to understand the gist of this change without having to read the entire JIRA / discussion.

gatorsmile committed:
[SPARK-13221] [SQL] Fixing GroupingSets when Aggregate Functions Containing GroupBy Columns

Using GroupingSets will generate a wrong result when Aggregate Functions containing GroupBy columns.

This PR is to fix it. Since the code changes are very small. Maybe we also can merge it to 1.6

For example, the following query returns a wrong result:
```scala
sql("select course, sum(earnings) as sum from courseSales group by course, earnings" +
     " grouping sets((), (course), (course, earnings))" +
     " order by course, sum").show()
```
Before the fix, the results are like
```
[null,null]
[Java,null]
[Java,20000.0]
[Java,30000.0]
[dotNET,null]
[dotNET,5000.0]
[dotNET,10000.0]
[dotNET,48000.0]
```
After the fix, the results become correct:
```
[null,113000.0]
[Java,20000.0]
[Java,30000.0]
[Java,50000.0]
[dotNET,5000.0]
[dotNET,10000.0]
[dotNET,48000.0]
[dotNET,63000.0]
```

UPDATE:  This PR also deprecated the external column: GROUPING__ID.

Author: gatorsmile <gatorsmile@gmail.com>

Closes #11100 from gatorsmile/groupingSets.

mgyucht committed:
Correct SparseVector.parse documentation

There's a small typo in the SparseVector.parse docstring (which says that it returns a DenseVector rather than a SparseVector), which seems to be incorrect.

Author: Miles Yucht <miles@databricks.com>

Closes #11213 from mgyucht/fix-sparsevector-docs.

BenFradet committed:
[SPARK-12247][ML][DOC] Documentation for spark.ml's ALS and collaborative filtering in general

This documents the implementation of ALS in `spark.ml` with example code in scala, java and python.

Author: BenFradet <benjamin.fradet@gmail.com>

Closes #10411 from BenFradet/SPARK-12247.

ueshin committed:
[SPARK-12976][SQL] Add LazilyGenerateOrdering and use it for RangePartitioner of Exchange.

Add `LazilyGenerateOrdering` to support generated ordering for `RangePartitioner` of `Exchange` instead of `InterpretedOrdering`.

Author: Takuya UESHIN <ueshin@happy-camper.st>

Closes #10894 from ueshin/issues/SPARK-12976.

Commit:
[SPARK-13280][STREAMING] Use a better logger name for FileBasedWriteAheadLog.

The new logger name is under the org.apache.spark namespace.
The detection of the caller name was also enhanced a bit to ignore
some common things that show up in the call stack.

Author: Marcelo Vanzin <vanzin@cloudera.com>

Closes #11165 from vanzin/SPARK-13280.

GayathriMurali commented:
@JoshRosen @zsxwing  I am currently trying to fix a null pointer exception. But on digging deeper, I see that there is no control over the daemon thread. It could start removing file objects even before it is added to the queue and there is no way to kill it. 
https://commons.apache.org/proper/commons-io/apidocs/src-html/org/apache/commons/io/FileCleaningTracker.html --- This implements a cleaner and more predictable Phantom Reference thread. I can write up a document on what needs to be changed to incorporate something like this, but can someone let me know if this approach would work.

JoshRosen committed:
[SPARK-13308] ManagedBuffers passed to OneToOneStreamManager need to be freed in non-error cases

ManagedBuffers that are passed to `OneToOneStreamManager.registerStream` need to be freed by the manager once it's done using them. However, the current code only frees them in certain error-cases and not during typical operation. This isn't a major problem today, but it will cause memory leaks after we implement better locking / pinning in the BlockManager (see #10705).

This patch modifies the relevant network code so that the ManagedBuffers are freed as soon as the messages containing them are processed by the lower-level Netty message sending code.

/cc zsxwing for review.

Author: Josh Rosen <joshrosen@databricks.com>

Closes #11193 from JoshRosen/add-missing-release-calls-in-network-layer.

Commit:
[SPARK-11627] Add initial input rate limit for spark streaming backpressure mechanism.

https://issues.apache.org/jira/browse/SPARK-11627

Spark Streaming backpressure mechanism has no initial input rate limit, it might cause OOM exception.
In the firest batch task ,receivers receive data at the maximum speed they can reach,it might exhaust executors memory resources. Add a initial input rate limit value can make sure the Streaming job execute  success in the first batch,then the backpressure mechanism can adjust receiving rate adaptively.

Author: junhao <junhao@mogujie.com>

Closes #9593 from junhaoMg/junhao-dev.

Commit:
[SPARK-13279] Remove O(n^2) operation from scheduler.

This commit removes an unnecessary duplicate check in addPendingTask that meant
that scheduling a task set took time proportional to (# tasks)^2.

Author: Sital Kedia <skedia@fb.com>

Closes #11175 from sitalkedia/fix_stuck_driver.

ueshin committed:
[SPARK-13357][SQL] Use generated projection and ordering for TakeOrderedAndProjectNode

`TakeOrderedAndProjectNode` should use generated projection and ordering like other `LocalNode`s.

Author: Takuya UESHIN <ueshin@happy-camper.st>

Closes #11230 from ueshin/issues/SPARK-13357.

chrisaycock committed:
[SPARK-13350][DOCS] Config doc updated to state that PYSPARK_PYTHON's default is "python2.7"

Author: Christopher C. Aycock <chris@chrisaycock.com>

Closes #11239 from chrisaycock/master.

jerryshao committed:
[SPARK-13109][BUILD] Fix SBT publishLocal issue

Add local ivy repo to the SBT build file to fix this.

Scaladoc compile error is fixed.

Author: jerryshao <sshao@hortonworks.com>

Closes #11001 from jerryshao/SPARK-13109.

shijinkui committed:
[SPARK-12953][EXAMPLES] RDDRelation writer set overwrite mode

https://issues.apache.org/jira/browse/SPARK-12953

fix error when run RDDRelation.main():
"path file:/Users/sjk/pair.parquet already exists"

Set DataFrameWriter's mode to SaveMode.Overwrite

Author: shijinkui <shijinkui666@163.com>

Closes #10864 from shijinkui/set_mode.

Commit:
[SPARK-13344][TEST] Fix harmless accumulator not found exceptions

See [JIRA](https://issues.apache.org/jira/browse/SPARK-13344) for more detail. This was caused by #10835.

Author: Andrew Or <andrew@databricks.com>

Closes #11222 from andrewor14/fix-test-accum-exceptions.

mengxr committed:
[MINOR][MLLIB] fix mllib compile warnings

This PR fixes some warnings found by `build/sbt mllib/test:compile`.

Author: Xiangrui Meng <meng@databricks.com>

Closes #11227 from mengxr/fix-mllib-warnings-201602.

srowen committed:
[SPARK-13324][CORE][BUILD] Update plugin, test, example dependencies for 2.x

Phase 1: update plugin versions, test dependencies, some example and third-party versions

Author: Sean Owen <sowen@cloudera.com>

Closes #11206 from srowen/SPARK-13324.

rxin committed:
Add github pull request template

srowen committed:
[SPARK-13371][CORE][STRING] TaskSetManager.dequeueSpeculativeTask compares Option and String directly.

## What changes were proposed in this pull request?

Fix some comparisons between unequal types that cause IJ warnings and in at least one case a likely bug (TaskSetManager)

## How was the this patch tested?

Running Jenkins tests

Author: Sean Owen <sowen@cloudera.com>

Closes #11253 from srowen/SPARK-13371.

GayathriMurali committed:
Redesigned the code to keep track of the creation and closing of the phantom reference thread.Fixed null pointer exceptions

GayathriMurali committed:
Redesigned the code to keep track of the creation and closing of the phantom reference thread.Fixed null pointer exceptions

SparkQA commented:
**[Test build #51502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51502/consoleFull)** for PR 11140 at commit [`23c853f`](https://github.com/apache/spark/commit/23c853f762ca995abf4702fc4881909ea88fd399).

SparkQA commented:
**[Test build #51502 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51502/consoleFull)** for PR 11140 at commit [`23c853f`](https://github.com/apache/spark/commit/23c853f762ca995abf4702fc4881909ea88fd399).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51502/
Test FAILed.
